lldb/source/Plugins/Language/CPlusPlus/LibStdcpp.cpp

```diff
@@ -254,13 +254,17 @@ bool lldb_private::formatters::LibStdcppStringSummaryProvider(
   } else
     addr_of_string =
         valobj.GetAddressOf(scalar_is_load_addr, &addr_type);
-  if (addr_of_string != LLDB_INVALID_ADDRESS) {
+
+  // We have to check for host address here
+  // because GetAddressOf returns INVALID for all non load addresses.
+  // But we can still format strings in host memory.
+  if (addr_of_string != LLDB_INVALID_ADDRESS ||
+        addr_type == eAddressTypeHost) {
```

Comment:
That I don't know. ValueObject does [explicitly return](https://github.com/llvm/llvm-project/blob/main/lldb/source/Core/ValueObject.cpp#L1408C12-L1408C32) `LLDB_INVALID_ADDRESS` when `addressType == eAddressTypeHost`. I thought this was weird and potentially returning the address of the in memory buffer made more sense, but that seemed like a major refactor for a minor string issue

Reply:
I think we need to be more careful here.  GetAddressOf is really meant to do "can you find an address in the target for this object".  We use it that way in a whole bunch of places, e.g.:

```
      cstr_address = GetAddressOf(true, &cstr_address_type);
    } else {
      // We have a pointer
      cstr_address = GetPointerValue(&cstr_address_type);
    }

    if (cstr_address == 0 || cstr_address == LLDB_INVALID_ADDRESS) {
      if (cstr_address_type == eAddressTypeHost && is_array) {
        const char *cstr = GetDataExtractor().PeekCStr(0);

```

So in that case we are expecting a host address type to return an invalid address from GetAddressOf.

This change worries me, I don't think it will be what other code expects?  Maybe you can get the value you need like the code above does instead?

Reply:
We could take the approach you provided. Greg and I originally talked and mentioned how `GetAddressOf` will return the pointer within a buffer in certain cases. I'm not enough of an expert here to have strong opinions but I think we're in a confusing middleground, where GetAddressOf works for Load addresses, but fails for a host address even if we have a pointer internally to a data buffer. 

I think solving that is more fundamental to my (2nd?) PR in lldb, but if we can all agree on the returning invaild for host addresses. I'm okay with that, as long as we also add a comment/documentation to the call that HostAddress addresses are invalid and to instead use a data extractor. My only opinion on that as a new contributor is the API to work with ValueObject's data is clunky, because I can seemingly only construct a copy via `GetData`.

Reply:
> I think we need to be more careful here. GetAddressOf is really meant to do "can you find an address in the target for this object". We use it that way in a whole bunch of places, e.g.:
> 
> ```
>       cstr_address = GetAddressOf(true, &cstr_address_type);
>     } else {
>       // We have a pointer
>       cstr_address = GetPointerValue(&cstr_address_type);
>     }
> 
>     if (cstr_address == 0 || cstr_address == LLDB_INVALID_ADDRESS) {
>       if (cstr_address_type == eAddressTypeHost && is_array) {
>         const char *cstr = GetDataExtractor().PeekCStr(0);
> ```
> 
> So in that case we are expecting a host address type to return an invalid address from GetAddressOf.
> 
> This change worries me, I don't think it will be what other code expects? Maybe you can get the value you need like the code above does instead?

The code above works, but I think it doesn't total sense from a readability standpoint. We have a `GetAddressOf` function that return the address kind and a value, I am not sure why we wouldn't want to take advantage of that and use it? The above code is a bit interesting as it makes a bunch of assumptions like "is this a host address and if this is an array, I know my data is in the data extractor at offset zero". I mean it works, but does it make sense to duplicate these kind of assumptions?

Make sense to fix GetAddressOf to take advantage of the API it is implementing. If the address kind can be host and we can return a valid host address value, I would say we use it. We will need to look over all uses of this API internally if we do change it.

Reply:
That's fine too, but you have to be careful here since getting host vrs. target address attribution wrong is one quick way to crash lldb.

Reply:
> Make sense to fix GetAddressOf to take advantage of the API it is implementing. If the address kind can be host and we can return a valid host address value, I would say we use it. We will need to look over all uses of this API internally if we do change it.

If we decide to go forward with this refactor, I think we should probably split this out into it's own independent PR and put this one on pause. As Jim mentioned above there are currently places that make assumptions based on the value object being host and getting back an invalid address, and we would need to correct for that.